tests/toxencryptsave/test_encryptsave.cpp

```diff
@@ -0,0 +1,648 @@
+{
+public:
+    static const char*  s_pass;
+    static const size_t s_pass_len;
+
+    uint8_t* passphrase;
+    size_t pplength;
+    TOX_PASS_KEY key;
+    TOX_ERR_KEY_DERIVATION key_derivation_error;
+
+    tox_derive_key_from_pass() : passphrase(NULL), pplength(), key(), key_derivation_error() {}
+    ~tox_derive_key_from_pass() {}
+
+    virtual void SetUp() {
+        pplength = s_pass_len;
+        passphrase = new uint8_t[ pplength ];
+        memmove(passphrase, s_pass, pplength);
+    }
+    virtual void TearDown() {
+        delete[] passphrase;
+    }
+};
+const char*  tox_derive_key_from_pass::s_pass = "12345678";
+const size_t tox_derive_key_from_pass::s_pass_len = strlen(s_pass);
+
+TEST_F(tox_derive_key_from_pass, all_args_are_good) {
+    EXPECT_TRUE( ::tox_derive_key_from_pass(passphrase, pplength, &key, &key_derivation_error) );
+    EXPECT_EQ(TOX_ERR_KEY_DERIVATION_OK, key_derivation_error);
+    EXPECT_TRUE( pass_was_zeroed(passphrase, pplength) );
+}
+
+TEST_F(tox_derive_key_from_pass, len_is_zero_passphrase_is_null) {
+    EXPECT_TRUE( ::tox_derive_key_from_pass(NULL, 0, &key, &key_derivation_error) );
+    EXPECT_EQ(TOX_ERR_KEY_DERIVATION_OK, key_derivation_error);
+    EXPECT_FALSE( pass_was_zeroed(passphrase, pplength) );
+}
+
+TEST_F(tox_derive_key_from_pass, len_is_zero_passphrase_is_not_null) {
+    EXPECT_TRUE( ::tox_derive_key_from_pass(passphrase, 0, &key, &key_derivation_error) );
+    EXPECT_EQ(TOX_ERR_KEY_DERIVATION_OK, key_derivation_error);
+    EXPECT_FALSE( pass_was_zeroed(passphrase, pplength) );
+}
+
+TEST_F(tox_derive_key_from_pass, error_is_null) {
+    EXPECT_TRUE( ::tox_derive_key_from_pass(passphrase, pplength, &key, NULL) );
+    EXPECT_EQ(TOX_ERR_KEY_DERIVATION_OK, key_derivation_error);
+    EXPECT_TRUE( pass_was_zeroed(passphrase, pplength) );
+}
+
+TEST_F(tox_derive_key_from_pass, passphrase_is_null) {
+    EXPECT_FALSE( ::tox_derive_key_from_pass(NULL, pplength, &key, &key_derivation_error) );
+    EXPECT_EQ(TOX_ERR_KEY_DERIVATION_NULL, key_derivation_error);
+    EXPECT_FALSE( pass_was_zeroed(passphrase, pplength) );
+}
+
+TEST_F(tox_derive_key_from_pass, key_is_null) {
+    EXPECT_FALSE( ::tox_derive_key_from_pass(passphrase, pplength, NULL, &key_derivation_error) );
+    EXPECT_EQ(TOX_ERR_KEY_DERIVATION_NULL, key_derivation_error);
+    EXPECT_TRUE( pass_was_zeroed(passphrase, pplength) );
+}
+
+
+class tox_pass_key_encrypt : public ::testing::Test
+{
+public:
+    static const char*  s_pass;
+    static const size_t s_pass_len;
+
+    static const uint8_t* s_data;
+    static const size_t   s_data_len;
+
+    static TOX_PASS_KEY s_key;
+
+    uint8_t* encrypted_out;
+    size_t   encrypted_len;
+    TOX_ERR_ENCRYPTION encryption_error;
+
+    tox_pass_key_encrypt() : encrypted_out(), encrypted_len(), encryption_error() {}
+    ~tox_pass_key_encrypt() {}
+
+    static void SetUpTestCase() {
+        uint8_t* passphrase = new uint8_t[ s_pass_len ];
```

Comment:
Well... do we REALLY need manual buffer management here? Use the `std::string` and you'll have easy buffers and easy init.